py-polars/polars/dataframe/frame.py

```diff
@@ -1306,6 +1306,12 @@ def __array__(self, dtype: Any = None) -> np.ndarray[Any, Any]:
         else:
             return self.to_numpy().__array__()
 
+    if sys.version_info >= (3, 12):
+
+        def __buffer__(self, flags: int) -> memoryview:
+            # see: https://peps.python.org/pep-0688/
+            return self.to_numpy().__buffer__(flags)
```

Comment:
Do we want to raise if not zero-copy? I assume the buffer protocol is meant to be like that.

Reply:
I think so too. Which means this would only work for numeric Series without nulls (and hardly ever for dataframes with multiple columns) on Python 3.12. With that in mind, it's probably better to leave this for now and properly implement this at a later point.

Reply:
Hmm, the plot thickens further! It seems that `numpy` preferentially calls `__buffer__` (if it exists) over `__array__` when running under Python 3.12 (I just confirmed this locally), eg:
```python
import numpy as np
import polars as pl
s_str = pl.Series( "a", ["aa", "bb", "cc", "dd"], dtype=pl.String )
np.char.capitalize( s_str )  # << calls `__buffer__` on 3.12
```
That would seem to imply that we may _have_ to match `__array__` (eg: zero_copy_only=False) on 3.12 or we'd break numpy interop by exposing a `__buffer__` that enforces zero_copy_only=True 

Given the niche application (3.12 only) and the potential side-effects, I think it makes sense to shelve this for now and revisit later 👌